### PR TITLE
Updated Homework #3 for 2022

### PR DESCRIPTION
**Pull request by borjasotomayor**
This is a major rewrite of Homework 3, which has been rewritten almost entirely from scratch. The homework now follows a tutorial-like approach to progressively building a Makefile for compiling a non-trivial program with multiple C files, and then asks the students to apply what they learned by writing a Makefile for a different codebase.

The current version is not yet complete, but the most important part (the tutorial explanation of Makefiles) is ready for review. The starter code can be found in this repository: https://github.com/uchicago-cmsc22000/hw3-starter-code-2022

P.S.- Sorry @mpcsh, but Obi Wan and General Grievous didn't survive the rewrite ;-(

### CODE REVIEW

**Comment by mpcsh** [comment]
😣😖😫😭

**Comment by borjasotomayor** [comment]
Ok, the homework writeup itself should be complete now (I've added the missing libgeometry and CMake sections). The only things missing are the rubric and the Gradescope autograder (so we can make sure that submitting the code works correctly). I'm hoping to have those done by the end of the day.

**Comment by borjasotomayor** [comment]
I have pushed a rubric for the homework. Since this homework is effectvely completely new, we're likely to encounter issues we did not anticipate, so I also welcome suggestions for other rubric items we may want to add.

File: website/hw/hw3_rubric.rst [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+
+Task 1: Makefile for the micro editor
+-------------------------------------
+
+- [Major] Not adding a ``Makefile`` in ``micro-editor/``
+- Having individual rules for each object file
+- Not using variables (``CC``, ``CFLAGS``, etc.)
+- Makefile does not correctly produce a ``micro`` executable.
+
+Task 2: Makefile for the micro editor
```
Comment (philipchenaux) [inline website/hw/hw3_rubric.rst:27]: I think this is meant to say Makefile for the librgraph library

**Review by philipchenaux — COMMENTED** [review]
No majors issues at all, works as expected. Included a couple of minor issues that I cam across.

File: website/hw/hw3.rst [hunk 2]
```diff
@@ -250,315 +313,879 @@ Once you’re done, add your ``Makefile`` to git, commit it, and push.
...
+Notice how we're compiling the Point and Polygon modules as we did
+before, but we are now using the ``-fPIC`` option. This flag tells
+gcc to enable *position-independent code*. Position-independence
+is beyond the scope of this homework, but it’s necessary for
+building shared libraries.
+
+Then, we simply link the ``point.o`` and ``polygon.o`` object files
+like we did before except, instead of producing an executable,
+we produce the ``libgeometry.so`` file (the ``-shared`` option
+instructs the linker to produce a library instead of an executable)
```
Comment (lilyehsani) [inline website/hw/hw3.rst:970]: Add . to end of sentence

**Review by lilyehsani — APPROVED** [review]
Looks great! I actually have a much better understanding of Makefiles now. I added a few comments about typos, but did not notice any ambiguities.

File: website/hw/hw3.rst [hunk 3]
```diff
@@ -250,315 +313,885 @@ Once you’re done, add your ``Makefile`` to git, commit it, and push.
...
+    ``micro-editor`` directory. You can ignore that file for now;
+    we will come back to it later in the homework.
+
+Building a library
+------------------
+
+So far, we've written a Makefile whose ultimate purpose is to build
+a ``polygon-area`` executable. However, let's say that we now
+wanted to build an additional executable, ``point-distance``, that
+relied on the Point module. Go ahead and create a file called
```
[the remark this answers is not in the document]
Comment (borjasotomayor) [reply website/hw/hw3.rst:911]: Good point, I've added a small note on this.

File: website/hw/hw3_rubric.rst (hunk 1 above)
Comment (borjasotomayor) [reply website/hw/hw3_rubric.rst]: Fixed

**Pull request merged**